### PR TITLE
Implement user teams, set API routing conventions, implement improvements from previous PR

### PR DESCRIPTION
**Pull request by JowiAoun**
This pull request serves to do the following, set to be more than the initial issue:
- Create Prisma team model
- Create team API route with team joining/leaving logic
- Implement team invitations using QR codes
- Create skeleton page for team invites: '/teams/join/{teamId}'
- Modify existing UserInformation model to enforce more non-nullable fields on the database level
- Set conventions in place for API routing (to be verified)
- Fix any additional warnings & errors, allowing successful production build
    - Attempts to deploy on AWS Amplify were successful, without containerized database

In addition, this request implements most of the development improvements suggested by Farabi in the previous PR.

### CODE REVIEW

File: src/app/_components/userProfile/UserProfile.tsx
Comment (HasithDeAlwis) [inline src/app/_components/userProfile/UserProfile.tsx:6]: My personal preference is that we keep this interface for code clarity

File: src/app/page.tsx
Comment (HasithDeAlwis) [inline src/app/page.tsx:8]: I don't think all const variables need to be upper case, especially on FE. Reserver upper case for const variables that aren't dynamic

Comment (HasithDeAlwis) [inline src/app/page.tsx:8]: Also, let's do this with `getServerSideProps()`. Refer to this [documentation](https://create.t3.gg/en/usage/next-auth#retrieving-session-server-side)

File: src/app/page.tsx
Comment (HasithDeAlwis) [inline src/app/page.tsx:1]: Let's move the content of this into a different folder. We'll discuss where exactly it should go in our meeting.

File: src/server/api/routers/team/procedures/getByTeamId.ts
Comment (HasithDeAlwis) [inline src/server/api/routers/team/procedures/getByTeamId.ts:6]: Thanks for implementing this early 🙏

File: src/server/auth.ts
Comment (HasithDeAlwis) [inline src/server/auth.ts:32]: dead code?

**Review by HasithDeAlwis — COMMENTED** [review]
Will continue reviewing this after the meeting.

File: src/app/page.tsx
Comment (ngc) [inline src/app/page.tsx:29]: I think it would be best if the UserProfile didn't take any props and just called getServerAuth session. Then we can wrap the whole UserProfile in a Suspense later. 

Let's try to keep data fetching within each respective component and avoid passing props unless necessary.

File: src/server/auth.ts
Comment (ngc) [inline src/server/auth.ts:70]: We should throw errors and log here instead of silently returning

File: src/server/api/trpc.ts
Comment (ngc) [inline src/server/api/trpc.ts:112]: nit: Can we get the actual type here instead of using ReturnType?

File: src/server/api/routers/team/team.ts
Comment (ngc) [inline src/server/api/routers/team/team.ts:14]: This is good to see. I like the doc comments!

File: src/app/_components/team/joinTeam.tsx
Comment (ngc) [inline src/app/_components/team/joinTeam.tsx:5]: Also, I'd rather not make this a separate component for now. Maybe just use the joinTeamAction callback in the component that uses this?

File: prisma/schema.prisma
Comment (ngc) [inline prisma/schema.prisma:70]: What was wrong here?

File: next.config.js
Comment (ngc) [inline next.config.js:14]: nicee

**Review by ngc — CHANGES_REQUESTED** [review]

File: src/app/_components/userProfile/UserProfile.tsx
Comment (MFarabi619) [reply src/app/_components/userProfile/UserProfile.tsx:6]: The reason I advocate for inline interfaces is that right now we don't have any re-use cases. Using inline interfaces reduces decision fatigue, keeps concerns separated, and keeps the data structures small.

Comment (HasithDeAlwis) [reply src/app/_components/userProfile/UserProfile.tsx:6]: Yea, that's fair, I understand your point. In my eyes, these interfaces wouldn't cause decision fatigue since they'd just be a standard that we have every time, and they'd help cleanup the code for function signatures, which can get messy without them. Could you elaborate on how it keeps data structures small?

File: src/server/auth.ts
Comment (JowiAoun) [reply src/server/auth.ts:32]: Build warning if left empty

File: prisma/migrations/20240721160053_create_team_model/migration.sql
Comment (JowiAoun) [reply prisma/migrations/20240721160053_create_team_model/migration.sql:24]: Not sure why Prisma created an index on the team name here. The only index on the Team model is on the ownerId.

File: src/server/api/trpc.ts
Comment (JowiAoun) [reply src/server/api/trpc.ts:112]: ReturnType simply gives us the type of the function here, which can vary as we modify it.
[Here](https://trpc.io/docs/server/context) is the documentation defining context type the same way

File: prisma/schema.prisma
Comment (JowiAoun) [reply prisma/schema.prisma:70]: Nothing modified here, most likely Prisma ORM refactoring extension

File: src/app/_components/team/joinTeam.tsx
Comment (JowiAoun) [reply src/app/_components/team/joinTeam.tsx:5]: Event handlers cannot be passed to client component props, Next docs recommend converting part of this (in this case the button) to a client component

File: src/app/_components/userProfile/UserProfile.tsx
Comment (MFarabi619) [reply src/app/_components/userProfile/UserProfile.tsx:6]: Forgot what was on my mind at the time I wrote the comment but I'll let you know once it comes back to me.

File: src/app/page.tsx
Comment (JowiAoun) [reply src/app/page.tsx]: Thanks for the review, but I am don't quite understand why we would want to do that. We are already fetching the session on the server

**Review by HasithDeAlwis — APPROVED** [review]

**Review by MFarabi619 — APPROVED** [review]

**Review by MFarabi619 — APPROVED** [review]

**Comment by MFarabi619** [comment]
@JowiAoun While you're waiting on Nathan's approval, please squash/amend commit groups such as these:

image
image

The images above are just examples, there are a number of commits that have the exact same message and/or make very minute changes that can be combined to create a much cleaner and more readable commit history.

**Comment by JowiAoun** [comment]
@MFarabi619 I'll begin doing all that from now on, unable to do it on this branch without making a fork because of merge commits

**Comment by MFarabi619** [comment]
> @MFarabi619 I'll begin doing all that from now on, unable to do it on this branch without making a fork because of merge commits

Sounds good thanks

File: src/app/_components/userProfile/UserProfile.tsx
Comment (ngc) [inline src/app/_components/userProfile/UserProfile.tsx:46]: 🔥??

Comment (MFarabi619) [reply src/app/_components/userProfile/UserProfile.tsx:46]: Man hit him with the question marks :rofl:

Comment (ngc) [inline src/app/_components/userProfile/UserProfile.tsx:46]: Also are these ids hashed? (i.e. do they look like '123' or are they 'xFjdgaor')? Because I think it's a problem if they aren't hashed. 

If they aren't hashed then someone could just iterate through all the teams and users and pull all their names just by scraping our site.

**Review by ngc — CHANGES_REQUESTED** [review]

File: src/app/_components/userProfile/UserProfile.tsx
Comment (JowiAoun) [reply src/app/_components/userProfile/UserProfile.tsx:46]: They're hashed 😉

https://github.com/cuhacking/hackathon/blob/df87210da03423c5d24ae8c2ede57f3c847e01c4/src/server/auth.ts#L60-L62

https://github.com/cuhacking/hackathon/blob/df87210da03423c5d24ae8c2ede57f3c847e01c4/src/server/auth.ts#L76-L82

Comment (JowiAoun) [reply src/app/_components/userProfile/UserProfile.tsx]: Made into separate component & used tailwind for styles

**Review by ngc — COMMENTED** [review]
Before this goes through could we get some more tests please?

**Comment by MFarabi619** [comment]
> Before this goes through could we get some more tests please?

@ngc What kind of tests?

**Pull request closed without merging**
